### ISSUE TITLE
Eject most Active Job adapters into gems·

### ISSUE THREAD

**Issue by dhh** [rails/rails#52929]
We used to ship with a million database adapters in the box, but eventually settled on just offering official support for mysql/pgsql/sqlite. Folks can of course use whatever database they desire, but the work to maintain db2 or sqlserver or oracle falls to those vendors or the broader community.

With Solid Queue becoming the default for Rails 8, it's time to do the same for Active Job. For starters, I think we can boil it down to the single-process adapters (inline/async/test) and then adapters for resque, sidekiq, and then of course Solid Queue (which actually just has the adapter in the gem!). Everything else should be extracted into separate adapter gems, like activejob-suckerpunch or whatever.

**Comment by byroot** [comment]
IMO this is welcome because that will get rid of quite a bit of dependencies in the rails Gemfile.

However before creating `activejob-*` gems, it would probably be worth reaching to the underlying gem maintainers to ask if they'd rather include the Active Job adapter in their gem.

IMO the list would be:

| Adapter                |  Task   | PR |
|------------------------|---------|----|
| backburner_adapter     | Extract |    |
| inline_adapter         | Keep    |    |
| resque_adapter         | Depends   |    |
| sneakers_adapter       | Extract |    |
| test_adapter           | Keep    |    |
| async_adapter          | Keep    |    |
| delayed_job_adapter    | Extract |    |
| queue_classic_adapter  | Extract |    |
| sidekiq_adapter        | Depends   |    |
| sucker_punch_adapter   | Extract |    |

- `Extract`: try to get it upstreamed and if not create a gem
- `Depends`: try to get it upstreamed but if not, keep.
- `Keep`: Should remain in core.

**Comment by dixpac** [comment]
I like it, and I’m open to helping with this feature.

> However before creating activejob-* gems, it would probably be worth reaching to the underlying gem maintainers to ask if they'd rather include the Active Job adapter in their gem.

👍

**Comment by dhh** [comment]
I'd love to get these adapters upstreamed. But maybe we can just do both. Would be nice to dump all these dependencies for 8.0. So maybe if upstreaming immediately isn't an option, we can at least put the long tail of adapters in a gem for staging for that to happen.

**Comment by byroot** [comment]
Is there really a rush? While as I said I'd love to reduce the dependencies needed to contribute to Rails, these doesn't appear actively harmful currently. So I don't really see why it would be seen as a blocker for Rails 8.0.

**Comment by dhh** [comment]
It's definitely not a blocker. But a nice to have if anyone is interested in doing the work before 8.0 👍

**Comment by dixpac** [comment]
@dhh @byroot wdyt as a first step I submit PR that removes sucker_punch adapter.
As far as I can see sucker_punch is pretty similar to `AsyncAdapter`, even the author [recommends this](https://github.com/brandonhilkert/sucker_punch?tab=readme-ov-file#faq)

> What is the difference between sucker_punch and ActiveJob::QueueAdapters::AsyncAdapter?
`Not much at this point. SuckerPunch existed before ActiveJob, but ultimately uses similar code under the covers. I'd recommend using AsyncAdapter if you're using Rails.`

**Comment by dhh** [comment]
Yeah we can actually just drop the sucker punch adapter outright imo.

**Comment by rafaelfranca** [comment]
Please! This is a thing I wish we had done early. Happy to see this moving.

**Comment by albus522** [comment]
Is there perhaps a way of keeping the test structure within rails and making it possible for outside adapters to run against a shared structure? This would better allow checking existing compatibility and adjusting to future changes.

**Comment by zzak** [comment]
@albus522 Each gem should just run their tests against Rails main so we can pick up breaking changes as early as possible. I've trimmed down the number of tests included in my PR to be adapter specific, but I think the queueing integration test will cover most things.

**Comment by Mdeliy** [comment]
![Uploading IMG_20241016_160644.jpg…]()

**Comment by morgoth** [comment]
To give overview of the current state:

- inline_adapter - kept
- test_adapter - kept
- async_adapter - kept
- backburner_adapter - upstreamed and deprecated (can be removed in Rails 9.0)
- sucker_punch_adapter -	removed in 8.1
- sidekiq_adapter - removed in 8.2
- sneakers_adapter - upstreamed and deprecated (can be removed in Rails 9.0)
- resque_adapter - upstreamed but not deprecated (can be removed in Rails 9.0)
- delayed_job_adapter - upstreamed and deprecated (can be removed in Rails 9.0)
- queue_classic_adapter - deprecated (can be removed in Rails 9.0)

**Comment by morgoth** [comment]
This can be closed. All of the adapters are deprecated. The 9.0 will have none of them.
